pyTigerGraph/schema.py

```diff
@@ -0,0 +1,268 @@
+        val_type = key_val.split(",")[1]
+        return Dict[_get_type(key_type), _get_type(val_type)]
+    elif attr_type == "BOOL":
+        return bool
+    elif attr_type == "datetime":
+        return datetime
+    else:
+        return attr_type
+
+
+def _py_to_tg_type(attr_type):
+    if attr_type == str:
+        return "STRING"
+    elif attr_type == int:
+        return "INT"
+    elif attr_type == float:
+        return "FLOAT"
+    elif attr_type == list:
+        raise TigerGraphException("Must define value type within list")
+    elif attr_type == dict:
+        raise TigerGraphException("Must define key and value types within dictionary/map")
+    elif attr_type == datetime:
+        return "DATETIME"
+    elif (str(type(attr_type)) == "<class 'typing._GenericAlias'>") and attr_type._name == "List":
+        val_type = _py_to_tg_type(attr_type.__args__[0])
+        if val_type.lower() in COLLECTION_VALUE_TYPES:
+            return "LIST<"+val_type+">"
+        else:
+            raise TigerGraphException(val_type + " not a valid type for the value type in LISTs.")
+    elif (str(type(attr_type)) == "<class 'typing._GenericAlias'>") and attr_type._name == "Dict":
+        key_type = _py_to_tg_type(attr_type.__args__[0])
+        val_type = _py_to_tg_type(attr_type.__args__[1])
+        if key_type.lower() in MAP_KEY_TYPES:
+            if val_type.lower() in COLLECTION_VALUE_TYPES:
+                return "MAP<"+key_type+","+val_type+">"
+            else:
+                raise TigerGraphException(val_type + " not a valid type for the value type in MAPs.")
+        else:
+            raise TigerGraphException(key_type + " not a valid type for the key type in MAPs.")
+    else:
+        print(attr_type)
+        if str(attr_type).lower() in BASE_TYPES:
+            return str(attr_type).upper()
+        else:
+            raise TigerGraphException(attr_type+"not a valid TigerGraph datatype.")
+
+
+
+@dataclass
+class Vertex(object):
+    attribute_edits = {"ADD": [], "DELETE": []}
+    incoming_edge_types = []
+    outgoing_edge_types = []
+
+    # TODO: I shouldn't have to run this, only use default value specified in dataclass
+    @classmethod
+    def define_primary_id(self, primary_id, primary_id_as_attribute):
+        self.primary_id = primary_id
+        self.primary_id_as_attribute = primary_id_as_attribute
+        self.primary_id_type = self.__annotations__[primary_id]
+        if not(_py_to_tg_type(self.primary_id_type).lower() in PRIMARY_ID_TYPES):
+            raise TigerGraphException(self.primary_id_type, "is not a supported type for primary IDs.")
+
+    @classmethod
+    def add_attribute(self, attribute, attribute_type):
+        for attr in self.attributes.values():
+            if attr == attribute:
+                raise TigerGraphException(attribute + " already exists as an attribute on "+self.vertex_type + " vertices")
+        self.attribute_edits["ADD"].append(attribute)
```

Comment:
Is attribute_type not used in this function?

Reply:
Removed, might need to re-instate it when I get to that logic